File: zig-util/packer.zig
// FalseTweaks
//
// Copyright (C) 2022-2025 FalsePattern
// All Rights Reserved
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, only version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

const std = @import("std");

pub fn main() !void {
    var allocator = std.heap.DebugAllocator(.{}).init;
    const alloc = allocator.allocator();
    var args = try std.process.argsWithAllocator(alloc);
    _ = args.next() orelse @panic("???");

    const cwd = std.fs.cwd();

    const output_file_path = args.next() orelse @panic("No output file");
    const output_file = try cwd.createFile(output_file_path, .{});
    defer output_file.close();
    var output_buf: [4096]u8 = undefined;
    var output_writer = output_file.writer(&output_buf);
    defer output_writer.interface.flush() catch @panic("Flush error");
    const writer = &output_writer.interface;
    var count: u32 = 0;
    try writer.writeInt(u32, 0, .big);

    while (true) {
        const name = args.next() orelse break;
        const file_path = args.next() orelse break;
        count += 1;

        try writer.writeInt(u8, @intCast(name.len), .big);
        try writer.writeAll(name);

        try writer.flush();

        const size_pos = output_writer.pos;

        try writer.writeInt(u32, 0, .big);

        var size: u32 = 0;

        {
            const input_file = try cwd.openFile(file_path, .{});
            defer input_file.close();
            var input_buf: [4096]u8 = undefined;
            var input_reader = input_file.reader(&input_buf);
            size = @intCast(try writer.sendFileAll(&input_reader, .unlimited));
        }
        try writer.flush();
        try output_writer.seekTo(size_pos);
        try writer.writeInt(u32, size, .big);
        try writer.flush();
        try output_writer.seekTo(try output_file.getEndPos());
    }

    try writer.flush();
    try output_writer.seekTo(0);
    try writer.writeInt(u32, count, .big);
}

File: src/main/java/com/falsepattern/falsetweaks/modules/animfix/stitching/packing2d/Packer.java
/*
 * This file is part of FalseTweaks.
 *
 * Copyright (C) 2022-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * FalseTweaks is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * FalseTweaks is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with FalseTweaks. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.falsetweaks.modules.animfix.stitching.packing2d;

import com.falsepattern.falsetweaks.modules.animfix.stitching.Rect2D;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public abstract class Packer<T extends Rect2D> {
    final int stripWidth;
    final List<T> rectangles;

    Packer(int stripWidth, List<T> rectangles) {
        this.stripWidth = stripWidth;
        this.rectangles = rectangles;
    }

    public static <U extends Rect2D> List<U> pack(List<U> rectangles, Algorithm algorithm, int stripWidth) {
        Packer<U> packer;
        switch (algorithm) {
            case FIRST_FIT_DECREASING_HEIGHT:
                packer = new PackerFFDH<>(stripWidth, rectangles);
                return packer.pack();
            case BEST_FIT_DECREASING_HEIGHT:
                packer = new PackerBFDH<>(stripWidth, rectangles);
                break;
            default:
                return new ArrayList<>();
        }
        return packer.pack();
    }

    public abstract List<T> pack();

    void sortByNonIncreasingHeight(List<T> rectangles) {
        rectangles.sort(Comparator.<T>comparingInt((rect) -> rect.height)
                                  .reversed());
    }
}
